### DISCUSSION TITLE
New way to readLines?

### DISCUSSION THREAD

**Discussion by alexgleason** [denoland/deno#23495] [Q&A] [answered]
[`readLines`](https://jsr.io/@std/io@0.223.0/doc/~/readLines) is deprecated, and says:

> (will be removed after 1.0.0) Use the {@link https://developer.mozilla.org/en-US/docs/Web/API/Streams_API | Web Streams API} instead.

But how do you use the Web Streams API with Deno.stdin and Deno.open?

Readlines is very convenient, and I haven't figured out how to update my code yet, eg:

```ts
async function* readStdin(): AsyncGenerator<InputMessage> {
  for await (const line of readLines(Deno.stdin)) {
    try {
      yield JSON.parse(line);
    } catch (e) {
      console.error(line);
      throw e;
    }
  }
}
```

**Comment by bartlomieju** [comment]
CC @iuioiua

**Comment by magurotuna** [comment, marked as answer]
Here's what I'd do when I want to parse jsonl file from the filesystem:

```ts
import { TextLineStream } from "jsr:@std/streams@0.223.0/text-line-stream";
import { JsonParseStream } from "jsr:@std/json@0.223.0/json-parse-stream";

using f = await Deno.open("./example.jsonl");
const readable = f.readable
  .pipeThrough(new TextDecoderStream()) // decode Uint8Array to string
  .pipeThrough(new TextLineStream()) // split string line by line
  .pipeThrough(new JsonParseStream()); // parse each chunk as JSON

for await (const data of readable) {
  console.log(data.wins);
}
```

Result:

```sh
❯ cat example.jsonl
───────┬────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: example.jsonl
───────┼────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ {"name": "Gilbert", "wins": [["straight", "7♣"], ["one pair", "10♥"]]}
   2   │ {"name": "Alexa", "wins": [["two pair", "4♠"], ["two pair", "9♠"]]}
   3   │ {"name": "May", "wins": []}
   4   │ {"name": "Deloise", "wins": [["three of a kind", "5♣"]]}
───────┴────────────────────────────────────────────────────────────────────────────────────────────────

❯ deno run --allow-read=. handle_jsonl.ts
[ [ "straight", "7♣" ], [ "one pair", "10♥" ] ]
[ [ "two pair", "4♠" ], [ "two pair", "9♠" ] ]
[]
[ [ "three of a kind", "5♣" ] ]
```

So basically idea here is that you can use `TextDecoderStream` and `TextLineStream` to achieve what `readLines` does.

**Reply by alexgleason** [reply]
Thank you @magurotuna! That worked for me. And now I understand how powerful the Stream API is.

Although, it's quote verbose for the common use-case of just reading lines. I would wrap it in a function. Maybe I would name the function `readLines`. And then I would consider adding it to the `@std/io` library. :joy:

**Reply by iuioiua** [reply]
Sorry for not getting back, @alexgleason, and thank you for answering, @magurotuna. Yep, the answer provided is correct. What's nice about the current `std/streams` package is that it allows for these modular, small utilities that make this code possible.

The key takeaway here is that we should improve our deprecation documentation. I've created https://github.com/denoland/deno_std/issues/4636 to track doing so.

**Reply by johndeighan** [reply]
You should specify where to get TextDecoderStream from. I found this works:

`import {TextDecoderStream} from "node:stream/web"
`

**Reply by alexgleason** [reply]
@johndeighan It's a global. You don't need to import it.
